articles/101_build_tool.md

```diff
@@ -0,0 +1,261 @@
+
+The tool needs to be able to build ROS 2 workspaces which can already be built using `ament_tools`.
+
+#### Build Gazebo including dependencies
+
+After cloning the repositories containing Gazebo and all its dependencies (excluding system packages) the tool needs to be able to build the set of packages.
+Meta information not inferable from the sources can be provided externally without adding or modifying any files in the workspace.
+After the build a single file can be sourced / invoked to setup the environment to use Gazebo (e.g. `GAZEBO_MODEL_PATH`).
+
+#### Mixing different build systems
+
+The build tool will support using different build systems within a single workspace.
+If these packages inter-operate with each other correctly depends also to a large degree on the build system.
+The build tool should ensure that it doesn't prevent that use case.
+
+### Software Criteria
+
+The tool aims to support a variety of build systems, use cases, and platforms.
+The above mentioned ones are mainly driven by the needs in the ROS ecosystem but the tool should also be usable outside the ROS ecosystem (e.g. for Gazebo).
+Therefore it should be designed in a way which enables extending its functionality.
+
+Assuming that the tool will be implemented in Python (since that is the case for existing tools) the entry point mechanism provides a convenient way to make the software extensible.
+Extensions don't even have to be integrated into the Python package containing the core logic of the build tool but can easily be provided by additional Python packages.
+This approach will not only foster a modular design and promote clear interfaces but enable external contributions without requiring them to be integrated in a single monolithic package.
+
+Several well known software principles apply:
+
+- Separation of concerns
+- Single Responsibility principle
+- Principle of Least Knowledge
+- Don’t repeat yourself
+- Keep it stupid simple
+- "Not paying for what you don't use"
+
+### Extension Points
+
+The following items are possible extension points to provide custom functionality:
+
+- contribute `verbs` to the command line tool (e.g. `build`, `test`)
+- contribute command line options for specific features (e.g. `build`, `test`)
+- discovery of packages (e.g. recursively crawling a workspace)
+- identification of packages and their meta information (e.g. from a `package.xml` file)
+- process a package (e.g. build a CMake package, test a Python package)
+- execution model (e.g. sequential processing, parallel processing)
+- output handling (e.g. console output, logfiles, status messages, notifications)
+- setup the environment (e.g. `sh`, `bash`, `bat`)
+- completion (e.g. `bash`, `Powershell`)
+
+## Possible Approaches
+
+In terms of flexibility neither of the existing build tools can already support the superset of features described in this article.
+There are two different paths possible to reach the goal of a universal build tool:
```

Comment:
What I suggested in a comment was a third way. Create a tool sitting on top of both ament_tools and catkin_tools, detecting the workspace nature, then invoking either of those tools. Basically treating catkin_tools and ament_tools as libraries (they can be forked as libraries without the CLI part). Possibly even enforce that catkin_tools as library is only used using Python2, and ament only with Python3. Once this runs, release it, then iteratively refactor moving common functionality to the universal layer. Should this approach be mentioned?